### ISSUE TITLE
Bug in loop for for pandas qcut

### ISSUE THREAD

**Issue by alandema** [spyder-ide/spyder#14984]
### What steps will reproduce the problem?

I have a list containing my dataframes, from which i iterate to binarize some columns. The loop, binarizes and change the values in column:

for k,i in enumerate(datasetlist):
    print(k)
    for j in range(0,len(i.columns)):
        if type(i.iloc[0,j])!=str:
            i.iloc[:,j]=pd.qcut(i.iloc[:,j],10,duplicates='drop')

After the loop, running "datasetlist[12]" gives me the anwer i'm expecting. But running the extensive name of the dataframe or view the dataframe in explorer, it is the old dataframe without changes.

## Versions

* Spyder version: 4.2.3 
* Python version: 3.8.8
* Qt version: 5.9.7
* PyQt5 version: 5.9.2
* Operating System: Windows 10

### Dependencies

```

# Mandatory:
atomicwrites >=1.2.0            :  1.4.0 (OK)
chardet >=2.0.0                 :  4.0.0 (OK)
cloudpickle >=0.5.0             :  1.6.0 (OK)
diff_match_patch >=20181111     :  20200713 (OK)
intervaltree >=3.0.2            :  3.1.0 (OK)
IPython >=7.6.0                 :  7.21.0 (OK)
jedi =0.17.2                    :  0.17.2 (OK)
jsonschema >=3.2.0              :  3.2.0 (OK)
keyring >=17.0.0                :  22.3.0 (OK)
nbconvert >=4.0                 :  6.0.7 (OK)
numpydoc >=0.6.0                :  1.1.0 (OK)
paramiko >=2.4.0                :  2.7.2 (OK)
parso =0.7.0                    :  0.7.0 (OK)
pexpect >=4.4.0                 :  4.8.0 (OK)
pickleshare >=0.4               :  0.7.5 (OK)
psutil >=5.3                    :  5.8.0 (OK)
pygments >=2.0                  :  2.8.1 (OK)
pylint >=1.0                    :  2.5.2 (OK)
pyls >=0.36.2;<1.0.0            :  0.36.2 (OK)
pyls_black >=0.4.6              :  0.4.6 (OK)
pyls_spyder >=0.3.2             :  0.3.2 (OK)
qdarkstyle >=2.8                :  2.8.1 (OK)
qtawesome >=0.5.7               :  1.0.2 (OK)
qtconsole >=5.0.1               :  5.0.2 (OK)
qtpy >=1.5.0                    :  1.9.0 (OK)
rtree >=0.8.3                   :  0.9.4 (OK)
setuptools >=39.0.0             :  52.0.0.post20210125 (OK)
sphinx >=0.6.6                  :  3.5.2 (OK)
spyder_kernels >=1.10.2;<1.11.0 :  1.10.2 (OK)
textdistance >=4.2.0            :  4.2.1 (OK)
three_merge >=0.1.1             :  0.1.1 (OK)
watchdog >=0.10.3;<2.0.0        :  1.0.2 (OK)
zmq >=17                        :  20.0.0 (OK)

# Optional:
cython >=0.21                   :  0.29.22 (OK)
matplotlib >=2.0.0              :  3.3.4 (OK)
numpy >=1.7                     :  1.20.1 (OK)
pandas >=1.1.1                  :  1.2.3 (OK)
scipy >=0.17.0                  :  1.6.1 (OK)
sympy >=0.7.3                   :  1.7.1 (OK)

# Spyder plugins:
spyder_notebook                 :  0.3.2 (OK)
```

**Comment by steff456** [comment]
Hi @alandema,

> But running the extensive name of the dataframe or view the dataframe in explorer, it is the old dataframe without changes.

Can you please post here screenshots of the dataframe before and after the procedure? I did not understand completely this sentence

**Comment by alandema** [comment]
Hello @steff456. Actually I saw that the issue is with an astype() conversion I was doing. If you want to look the code, it follows below. But I handled the issue excluding the cmc=cmc.astype(str) line and converting only the columns that are in fact str.
`
import pandas as pd
import numpy as np
from sklearn.metrics import f1_score
from scipy.stats import ttest_ind
from sklearn.preprocessing import OrdinalEncoder
'''
ONE TIME RUN SCRIPT. AS I KEEP REUSING THE SAME VARIABLE NAMES, IT MAY CAUSE ERROR
WHEN RERUNNING THE SCRIPT WITHOUT CLEANING THE VARIABLES
'''


'''
THE MATH PART (NEED WORK)

#As paper "Bayesian belief..." chapter 4.1
def UPTAN(training, attr, classes):
    return

#As paper "Bayesian Belief" chapter 5 (can look at paper "a rule-based...")
def isert_uncertain(u,[attri]):
    return

'''

'''
LOADING DATASET
'''
vote=pd.read_csv(r'house-votes-84.csv',header=None)
balloons1=pd.read_csv(r'yellow-small+adult-stretch.csv',header=None)
balloons2=pd.read_csv(r'adult-stretch.csv',header=None)
balloons3=pd.read_csv(r'yellow-small.csv',header=None)
balloons4=pd.read_csv(r'adult+stretch.csv',header=None)
balloons=pd.concat([balloons1,balloons2,balloons3,balloons4])
br_ca=pd.read_csv(r'breast-cancer.csv',header=None)
br_ca_win=pd.read_csv(r'breast-cancer-wisconsin.csv',header=None,sep=';')
car=pd.read_csv(r'car.csv',header=None)
flare=pd.read_csv(r'flare1csv.csv',header=None,sep=';')
tictac=pd.read_csv(r'tic-tac-toe.csv',header=None)
zoo=pd.read_csv(r'zoo.csv',header=None)
post_op=pd.read_csv(r'post-operative.csv',header=None)
nursery=pd.read_csv(r'nursery.csv',header=None)
krkp=pd.read_csv(r'kr-vs-kp.csv',header=None)
mushroom=pd.read_csv(r'agaricus-lepiota.csv',header=None)
cmc=pd.read_csv(r'cmc.csv',header=None)
creditcard=pd.read_csv(r'crx.csv',header=None)
glass=pd.read_csv(r'glass.csv',header=None)
heart=pd.read_csv(r'processed.cleveland.csv',header=None)
liver=pd.read_csv(r'bupa.csv',header=None)
ecoli=pd.read_csv(r'ecoli.csv',header=None,sep=';')
iris=pd.read_csv(r'iris.csv',header=None)
hayes=pd.read_csv(r'hayes-roth.csv',header=None)

'''
SPLITTING THE CLASS COLUMNS AND TREATING ANOTHER UNUSEFULL COLUMNS
'''
vote_class=balloons[0]
vote.drop([0],axis=1, inplace=True)
balloons_class=balloons[4]
balloons.drop([4],axis=1, inplace=True)
br_ca_class=br_ca[[0]]
br_ca.drop([0],axis=1, inplace=True)
br_ca_win_class=br_ca_win[10]
br_ca_win.drop([0,10],axis=1, inplace=True)
car_class=car[6]
car.drop([6],axis=1, inplace=True)
flare_class=flare[[10,11,12]]
flare.drop([10,11,12],axis=1, inplace=True)
hayes_class=hayes[[5]]
hayes.drop([0,5],axis=1, inplace=True)
iris_class=iris[[4]]
iris.drop([4],axis=1, inplace=True)
ecoli_class=ecoli[[8]]
ecoli.drop([0,8],axis=1, inplace=True)
liver_class=liver[[6]]
liver.drop([6],axis=1, inplace=True)
heart_class=heart[[13]]
heart.drop([13],axis=1, inplace=True)
glass_class=glass[[10]]
glass.drop([0,10],axis=1, inplace=True)
creditcard_class=creditcard[[15]]
creditcard.drop([15],axis=1, inplace=True)
creditcard.iloc[:,1].replace('?',9,inplace=True)
cmc_class=cmc[[9]]
cmc.drop([9],axis=1, inplace=True)
mushroom_class=mushroom[[0]]
mushroom.drop([0],axis=1, inplace=True)
krkp_class=krkp[[36]]
krkp.drop([36],axis=1, inplace=True)
nursery_class=nursery[[8]]
nursery.drop([8],axis=1, inplace=True)
post_op_class=post_op[[8]]
post_op.drop([8],axis=1, inplace=True)
zoo_class=zoo[[0]]
zoo.drop([0],axis=1, inplace=True)
tictac_class=tictac[[9]]
tictac.drop([9],axis=1, inplace=True)

'''
PROCESSING MISSING DATA VALUES
'''
datasetlist=[vote,balloons,br_ca,br_ca_win,car,flare,tictac,zoo,post_op,nursery,krkp,mushroom,cmc,creditcard,glass,
             heart,liver,ecoli,iris,hayes]
for i in datasetlist:
    for j in range(0,len(i.columns)):
        repl=i.iloc[:,j].value_counts().idxmax()
        
        i.iloc[:,j].replace('?',repl,inplace=True)

'''
MODIFYING THE COLUMNS TYPE AS DISCRETE OR CONTINOUS
'''
br_ca_win=br_ca_win.astype(str)
br_ca[6]=br_ca[6].astype(str)
cmc=cmc.astype(str)
cmc[[0,3]]=cmc[[0,3]].astype(int)
creditcard[[1,13]]=creditcard[[1,13]].astype(float)
flare=flare.astype(str)
hayes=hayes.astype(str)
heart[[1,2,5,6,8,10,11,12]]=heart[[1,2,5,6,8,10,11,12]].astype(str)
zoo=zoo.astype(str)

'''
BINNING CONTINOUS DATA
'''
for k,i in enumerate(datasetlist):
    for j in range(0,len(i.columns)):
        if type(i.iloc[0,j])!=str:
            i.iloc[:,j]=pd.qcut(i.iloc[:,j],10,duplicates='drop')
            
cmc
datasetlist[12]`

**cmc is different from datasetlist[12]**

**Comment by steff456** [comment]
So did you fix your issue @alandema ?

**Comment by alandema** [comment]
> So did you fix your issue @alandema ?

Yes.

**Comment by steff456** [comment]
Perfect! I'm closing this issue then.

Thanks for reporting!
